### PR TITLE
Analyze Sepsis to TB Domain Adaptation

### PR DESCRIPTION
**Pull request by ben-heil**
Main Ideas:
- Are models trained on sepsis data able differentiate between tb and non-tb gene expression?
- Add the ability to pull out data to function as a test set based on a single label e.g. 'tuberculosis'. This prevents an issue where data was leaking from the training set to the test set caused by all healthy samples' gene expression being used to train the original model.

Changes:
- Create the notebook `1.Transfer_Sepsis_to_TB.ipynb`
- Modify the main function in `classifier.py` to use the new `get_data_dirs` and `train_tune_split` functions
- Increase cross-platform compatibility of dataset.py by changing from manual path parsing to `os.path`
- Create the `get_data_dirs` utility function to extract all expression data containing subdirs from the data directory
- Implement `extract_test_dirs`, a function that splits directories (corresponding to studies) into those containing the label passed in and those without. For example, this function is used in `1.Transfer_Sepsis_to_TB.ipynb` to hold out all the healthy and unhealthy samples from studies on tuberculosis.

### CODE REVIEW

File: analyses/1.Transfer_Sepsis_to_TB.ipynb [hunk 1]
```diff
@@ -0,0 +1,425 @@
...
+   "source": [
+    "irm_results = classifier.train_with_irm(net, map_file, train_dirs, tune_dirs, gene_file, num_epochs, \n",
+    "                                        loss_scaling_factor, label_to_encoding, device, logger, '../logs/sepsis_irm.pkl', 100)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Print tuning set accuracy at the end of training for ERM and IRM models"
```
Comment (jjc2718) [inline analyses/1.Transfer_Sepsis_to_TB.ipynb:188]: Interesting. Did you expect IRM to have lower accuracy on the tuning set here?

File: analyses/1.Transfer_Sepsis_to_TB.ipynb [hunk 2]
```diff
@@ -0,0 +1,425 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 18,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "ERM trained network tb accuracy: 0.7375964718853363\n",
```
Comment (jjc2718) [inline analyses/1.Transfer_Sepsis_to_TB.ipynb:327]: Interesting that IRM seems to have a better (lower) loss value on the TB data, but the accuracy is quite a bit lower.

File: analyses/1.Transfer_Sepsis_to_TB.ipynb [hunk 3]
```diff
@@ -0,0 +1,425 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Conclusion\n",
+    "\n",
+    "While they don't perform as well on the transferred domain as the source domain, both the erm and irm trained networks have better than random (and better than the frequency of the more common class). Interestingly, the ERM trained model predicted tuberculosis 8/9 of the time, while the IRM trained model predicted tb about 5/9 of the time. The true frequency of tb in the dataset is 77%.\n",
+    "\n",
+    "These results are evidence that domain adaptation works to some extent on gene expression data even between diseases."
```
Comment (jjc2718) [inline analyses/1.Transfer_Sepsis_to_TB.ipynb:400]: Neat! I agree with your conclusions.

It's been a while since I looked at your code so I don't remember - are you doing anything (e.g. reweighting your loss function) to correct for class imbalance? You could try using the `pos_weight` argument to `BCEWithLogitsLoss` to do this (example from my code [here](https://github.com/greenelab/netreg/blob/fc3443a59601071dceb8e174130de524cf4dd0d3/utilities/classify_pytorch.py#L171)).

**Review by jjc2718 — APPROVED** [review]
Looks good! Just a few comments/observations on your results.

File: analyses/1.Transfer_Sepsis_to_TB.ipynb (hunk 1 above)
Comment (ben-heil) [reply analyses/1.Transfer_Sepsis_to_TB.ipynb:188]: I wasn't sure what to expect, because I'm still not sure where the models fall on the overfitting/underfitting scale. IRM acts as a form of regularization, so I guess it makes sense that the performance would be worse since the models are only working with a few thousand samples.

File: analyses/1.Transfer_Sepsis_to_TB.ipynb (hunk 2 above)
Comment (ben-heil) [reply analyses/1.Transfer_Sepsis_to_TB.ipynb:327]: This was surprising to me too, especially because ERM optimizes the loss function we're using as the metric here while IRM jointly optimizes the BCE loss and the gradient penalty. 

The ERM model is predicting mainly the dominant class, while the IRM class predictions are more balanced. This lowers the accuracy in IRM, but the decreased loss might indicate that IRM's representation is more biologically interesting

File: analyses/1.Transfer_Sepsis_to_TB.ipynb (hunk 3 above)
Comment (ben-heil) [reply analyses/1.Transfer_Sepsis_to_TB.ipynb:400]: Yep, since there's a big class imbalance I'm adjusting pos_weight to account for it. The way I'm doing this is slightly different in IRM and ERM though, which may contribute to the different predicted class ratios when doing domain adaptation.
